libcrux-ml-kem/src/hash_functions.rs

```diff
@@ -54,76 +54,112 @@ pub(crate) mod portable {
     ///
     /// It's only used for SHAKE128.
     /// All other functions don't actually use any members.
+    #[cfg_attr(hax, hax_lib_macros::opaque_type)]
     pub(crate) struct PortableHash<const K: usize> {
         shake128_state: [KeccakState; K],
     }
 
+    #[inline(always)]
+    fn G(input: &[u8]) -> [u8; G_DIGEST_SIZE] {
+        let mut digest = [0u8; G_DIGEST_SIZE];
+        portable::sha512(&mut digest, input);
+        digest
+    }
+
+    #[inline(always)]
+    fn H(input: &[u8]) -> [u8; H_DIGEST_SIZE] {
+        let mut digest = [0u8; H_DIGEST_SIZE];
+        portable::sha256(&mut digest, input);
+        digest
+    }
+
+    #[inline(always)]
+    fn PRF<const LEN: usize>(input: &[u8]) -> [u8; LEN] {
+        let mut digest = [0u8; LEN];
+        portable::shake256(&mut digest, input);
+        digest
+    }
+
+    #[inline(always)]
+    fn PRFxN<const K:usize, const LEN: usize>(input: &[[u8; 33]; K]) -> [[u8; LEN]; K] {
+        debug_assert!(K == 2 || K == 3 || K == 4);
+
+        let mut out = [[0u8; LEN]; K];
+        for i in 0..K {
+            portable::shake256(&mut out[i], &input[i]);
+        }
+        out
+    }
+
+    #[inline(always)]
+    fn shake128_init_absorb<const K:usize>(input: [[u8; 34]; K]) -> PortableHash<K> {
+        debug_assert!(K == 2 || K == 3 || K == 4);
+
+        let mut shake128_state = [shake128_init(); K];
+        for i in 0..K {
+            shake128_absorb_final(&mut shake128_state[i], &input[i]);
+        }
+        PortableHash {
+            shake128_state,
+        }
+    }
+
+    #[inline(always)]
+    fn shake128_squeeze_three_blocks<const K:usize>(st: &mut PortableHash<K>) -> [[u8; THREE_BLOCKS]; K] {
+        debug_assert!(K == 2 || K == 3 || K == 4);
+
+        let mut out = [[0u8; THREE_BLOCKS]; K];
+        for i in 0..K {
+            shake128_squeeze_first_three_blocks(&mut st.shake128_state[i], &mut out[i]);
+        }
+        out
+    }
+
+    #[inline(always)]
+    fn shake128_squeeze_block<const K:usize>(st: &mut PortableHash<K>) -> [[u8; BLOCK_SIZE]; K] {
+        debug_assert!(K == 2 || K == 3 || K == 4);
+
+        let mut out = [[0u8; BLOCK_SIZE]; K];
+        for i in 0..K {
+            shake128_squeeze_next_block(&mut st.shake128_state[i], &mut out[i]);
+        }
+        out
+    }
+
     impl<const K: usize> Hash<K> for PortableHash<K> {
         #[inline(always)]
         fn G(input: &[u8]) -> [u8; G_DIGEST_SIZE] {
-            let mut digest = [0u8; G_DIGEST_SIZE];
-            portable::sha512(&mut digest, input);
-            digest
+            G(input)
```

Comment:
Added a link to Issue https://github.com/hacspec/hax/issues/616  at the top of the file.
Extracting the functions out of the "impl" is needed both for abstraction in F* (interface hiding implementation details) and to make the F* typechecker able to handle this code.
However, this could be done automatically in hax without needing to uglify the Rust, hence the issue above.